core/relayer/deliver.go

```diff
@@ -0,0 +1,201 @@
+		if err != nil && strings.Contains(err.Error(), "only once") {
+			tmpDeliverMeta := &relayerpb.DeliverMeta{
+				LastDeliverBlockHeight: cmd.AnchorBlockHeight,
+			}
+			updateErr := cmd.UpdateDeliverMeta(tmpDeliverMeta)
+			if updateErr != nil {
+				panic("update deliver meta failed, err:" + updateErr.Error())
+			}
+			cmd.meta = &relayerpb.DeliverMeta{
+				LastDeliverBlockHeight: cmd.AnchorBlockHeight,
+			}
+			return err
+		} else if err != nil {
+			fmt.Println("[deliver] preExe for synchronzing anchor block failed, err:", err)
+			time.Sleep(time.Duration(1) * time.Second)
+			continue
+		}
+		txid, err := cmd.SendTx(tx)
+		if err != nil && strings.Contains(err.Error(), "only once") {
+			tmpDeliverMeta := &relayerpb.DeliverMeta{
+				LastDeliverBlockHeight: cmd.AnchorBlockHeight,
+			}
+			updateErr := cmd.UpdateDeliverMeta(tmpDeliverMeta)
+			if updateErr != nil {
+				panic("update deliver meta failed, err:" + updateErr.Error())
+			}
+			cmd.meta = &relayerpb.DeliverMeta{
+				LastDeliverBlockHeight: cmd.AnchorBlockHeight,
+			}
+			return err
+		} else if err != nil {
+			fmt.Println("[deliver] preExe for synchronzing anchor block failed, err:", err)
+			time.Sleep(time.Duration(1) * time.Second)
+			continue
+		}
+		fmt.Println("[deliver] txid:", txid)
+		// 更新DeliverMeta
+		tmpDeliverMeta := &relayerpb.DeliverMeta{
+			LastDeliverBlockHeight: cmd.AnchorBlockHeight,
+		}
+		updateErr := cmd.UpdateDeliverMeta(tmpDeliverMeta)
+		if updateErr != nil {
+			panic("update deliver meta failed, err:" + updateErr.Error())
+		}
+		cmd.meta = &relayerpb.DeliverMeta{
+			LastDeliverBlockHeight: cmd.AnchorBlockHeight,
+		}
+
+		break
+	}
+
+	return nil
+}
+
+func (cmd *DeliverBlockCommand) DeliverBlockHeader() error {
+	// 持续获取本地区块头数据并转发给目标链的区块头合约
+	currHeight := cmd.meta.GetLastDeliverBlockHeight() + 1
+	for {
+		if currHeight+3 >= cmd.QueryMeta.GetLastQueryBlockHeight() {
+			// 需要等待几个区块再Deliver
+			fmt.Println("[deliver] should wait for seconds to deliver again", "currHeight:", currHeight, " deliverBlockHeight:", cmd.QueryMeta.GetLastQueryBlockHeight())
+			time.Sleep(time.Duration(1) * time.Second)
+			continue
+		}
+		// 按照高度获取本地区块头
+		blockBuf, getErr := cmd.Storage.GetBlockHeaderByHeight(currHeight)
+		// 本地没有该区块头, 稍等一会再重试
+		if common.NormalizedKVError(getErr) == common.ErrKVNotFound {
+			fmt.Println("[deliver] block is not found in the storage, err:", getErr)
+			time.Sleep(time.Duration(1) * time.Second)
+			continue
+		} else if getErr != nil {
+			panic("gey block failed, err:" + getErr.Error())
+		}
+		// prepare para for pre-exe
+		args := make(map[string][]byte)
+		args["blockHeader"] = blockBuf
+		// set preExe parameter
+		moduleName := cmd.Cfg.ContractConfig.ModuleName
+		contractName := cmd.Cfg.ContractConfig.ContractName
+		methodName := cmd.Cfg.ContractConfig.UpdateMethod
+		tx, err := cmd.PreExe(moduleName, contractName, methodName, args)
+		if err != nil {
+			if strings.Contains(err.Error(), "missing preHash") {
```

Comment:
keep error strings as Constants to avoid duplicate hard-code strings everywhere.

Reply:
Corrected